user_tools/src/spark_rapids_pytools/resources/dev/prepackage_mgr.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module defines the logic to fetch all dependencies and resources required to run the user-tools
+without need to access the web during runtime.
+"""
+
+import os
+import shutil
+import tarfile
+from concurrent.futures import ThreadPoolExecutor
+from typing import Optional
+
+import fire
+
+from spark_rapids_pytools.cloud_api.sp_types import CloudPlatform
+from spark_rapids_pytools.common.prop_manager import JSONPropertiesContainer
+from spark_rapids_pytools.common.sys_storage import FSUtil
+from spark_rapids_pytools.common.utilities import Utils
+
+
+# Defines the constants and static configurations
+prepackage_conf = {
+    '_supported_platforms': [csp.value for csp in CloudPlatform if csp != CloudPlatform.NONE],
+    '_configs_suffix': '-configs.json',
+    '_mvn_base_url': 'https://repo1.maven.org/maven2/com/nvidia/rapids-4-spark-tools_2.12',
+    '_folder_name': 'csp-resources'
+}
+
+
+class PrepackageMgr:   # pylint: disable=too-few-public-methods
+    """ Class that handles downloading dependencies to be pre-packaged ahead of runtime.
+
+    The script can be triggered by passing the arguments and the properties of the class
+    For example:
+
+    $> python prepackage_mgr.py --resource_dir=RESOURCE_DIR --tools_jar=TOOLS_JAR run
+
+    :param resource_dir: root folder where the configuration files are located.
+    :param dest_dir: directory in which the resources are downloaded and stored.
+           If missing, the tool creates a subfolder '$resource_dir/csp-resources'.
+           Warning: the 'dest_dir' may be deleted when 'archive_enabled' is set to True.
+    :param tools_jar: path of the spark-rapids-user-tools jar file. Typically, this is the snapshot
+           jar file of the current build.
+    :param archive_enabled: a flag to enable/disable compressing the resources.
+           Note that a wheel-package with compressed prepackaged resources is 30% less size
+           compared to the non-compressed one (~ 600 Vs. 900 MB).
+           When enabled, the prepackaged-resources are stored in '$resource_dir/csp-resources.tgz'.
+           If the 'dest_dir' is provided, then the output is stored as '$dest_dir/../csp-resources.tgz'
```

Comment:
I am a bit confused on this. If `archive_enabled` is set to `True` and `dest_dir` is provided, then will the $dest_dir be deleted but  csp-resources.tgz will be saved in parent directory of `dest_dir`?

Reply:
mmm, yes. to be consistent with the default behavior.
The default is to add `/csp-resources` subdirectory to the resources. Then, the tgz file will be `/resources/csp-resources.tgz`; but not `/resources/csp-resources/csp-resources.tgz`

Anyway, That's not going to be used by the build-script.
Providing `dest_dir` is more like a debugging/testing purpose. For example, to test the script separately without polluting the `/resources` folder.
The warning means that the user should not provide a dest-directory that includes some other files, or she/he may loses them.